File: 98.rb
# https://leetcode.com/problems/validate-binary-search-tree/
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Boolean}
def inorder(node, res=[])
	return if !node

	inorder(node.left, res)
	res << node.val
	inorder(node.right, res)
	res
end

def is_valid_bst(root)
	return true if !root

	res = inorder(root)
	res.compact!
	res == res.uniq.sort
end

File: js/98.js
// https://leetcode.com/problems/validate-binary-search-tree/
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isValidBST = function(root) {
	if (!root) {
		return true
	}
	let curr
	let inorder = (root) => {
		if (!root) {
			return true
		}
		let isLeftValid = inorder(root.left)
		if (curr === undefined) {
			curr = root.val
		} else {
			if (curr >= root.val) {
				return false
			} else {
				curr = root.val
			}
		}
		let isRightValid = inorder(root.right)
		return isLeftValid && isRightValid
	}
	return inorder(root)
};
